src/Workspaces/Remote/Core/SolutionAssetProvider.cs

```diff
@@ -62,171 +60,20 @@ async ValueTask WriteAssetsSuppressedFlowAsync(PipeWriter pipeWriter, Checksum s
             }
         }
 
-        private async ValueTask WriteAssetsWorkerAsync(
+        private ValueTask WriteAssetsWorkerAsync(
             PipeWriter pipeWriter,
             Checksum solutionChecksum,
             AssetPath assetPath,
             ReadOnlyMemory<Checksum> checksums,
             CancellationToken cancellationToken)
         {
             var assetStorage = _services.GetRequiredService<ISolutionAssetStorageProvider>().AssetStorage;
-            var serializer = _services.GetRequiredService<ISerializerService>();
             var scope = assetStorage.GetScope(solutionChecksum);
 
-            using var _ = Creator.CreateResultMap(out var resultMap);
-
-            await scope.AddAssetsAsync(assetPath, checksums, resultMap, cancellationToken).ConfigureAwait(false);
-
-            cancellationToken.ThrowIfCancellationRequested();
-
-            using var stream = new PipeWriterStream(pipeWriter);
-            await RemoteHostAssetSerialization.WriteDataAsync(
-                stream, resultMap, serializer, scope.ReplicationContext,
-                solutionChecksum, checksums, cancellationToken).ConfigureAwait(false);
-        }
-
-        /// <summary>
-        /// Simple port of
-        /// https://github.com/AArnott/Nerdbank.Streams/blob/dafeb5846702bc29e261c9ddf60f42feae01654c/src/Nerdbank.Streams/BufferWriterStream.cs#L16.
-        /// Wraps a <see cref="PipeWriter"/> in a <see cref="Stream"/> interface.  Preferred over <see
-        /// cref="PipeWriter.AsStream(bool)"/> as that API produces a stream that will synchronously flush after
-        /// <em>every</em> write.  That's undesirable as that will then block a thread pool thread on the actual
-        /// asynchronous flush call to the underlying PipeWriter
-        /// </summary>
-        /// <remarks>
-        /// Note: this stream does not have to <see cref="PipeWriter.Complete"/> the underlying <see cref="_writer"/> it
-        /// is holding onto (including within <see cref="Flush"/>, <see cref="FlushAsync"/>, or <see cref="Dispose"/>).
-        /// Responsibility for that is solely in the hands of <see cref="WriteAssetsAsync"/>.
-        /// </remarks>
-        private class PipeWriterStream : Stream, IDisposableObservable
```

Comment:
no need for this specizlied stream.  we own all the async parts of working with the stream ourselves.  and the synchronous parts go to a different in-memory stream where sync over async is not a concern.